### PR TITLE
HttpResponseHeaders.CopyToFast move directly to next rather than `if` branching

### PR DESCRIPTION
**Pull request by benaadams**
Can use `BitOperations.TrailingZeroCount` on the set bits to determine which is the next header; rather than testing each bit in sequence via fall-through, droping x37 `if` statements from the `CopyToFast` response header output method.

Second commit is pulling `Content-Length` to the first header in tests (as its a special `long?` outlier header)

from
```csharp
case 5: // Header: "Accept-Ranges"
    if ((tempBits & 0x10L) != 0) // Test
    {
        tempBits ^= 0x10L;
        values = ref _headers._AcceptRanges;
        keyStart = 48;
        keyLength = 17;
        next = 6;
        break; // OutputHeader
    }
    goto case 6; // Fall-through

case 6: // Header: "Access-Control-Allow-Credentials"
    if ((tempBits & 0x20L) != 0) // Test
    {
        tempBits ^= 0x20L;
        values = ref _headers._AccessControlAllowCredentials;
        keyStart = 65;
        keyLength = 36;
        next = 7;
        break; // OutputHeader
    }
    goto case 7; // Fall-through
```
to
```csharp
case 4: // Header: "Accept-Ranges"
    values = ref _headers._AcceptRanges;
    keyStart = 48;
    keyLength = 17;
    break; // OutputHeader

case 5: // Header: "Access-Control-Allow-Credentials"
    values = ref _headers._AccessControlAllowCredentials;
    keyStart = 65;
    keyLength = 36;
    break; // OutputHeader
```
735 bytes less IL
```diff
- // IL Code size 2325 (0x915)
+ // IL Code size 1590 (0x636)
```
871 bytes less asm
```diff
- ; Total bytes of asm code: 3910
+ ; Total bytes of asm code: 3039
```

| Branch |                 Type |      Mean |         Op/s |  Delta |
|------- |--------------------- |----------:|-------------:|-------:|
|   main | TechEmpowerPlaintext |  56.79 ns | 17,609,767.6 |      - |
|     PR | TechEmpowerPlaintext |  53.53 ns | 18,679,950.2 |  +6.1% |
|   main |     PlaintextChunked |  61.89 ns | 16,158,595.4 |      - |
|     PR |     PlaintextChunked |  47.82 ns | 20,912,547.5 | +29.4% |
|   main |  PlaintextWithCookie | 133.94 ns |  7,466,115.3 |      - |
|     PR |  PlaintextWithCookie | 126.26 ns |  7,920,045.3 |  +6.1% |
|   main | Plain(...)ookie [26] | 163.29 ns |  6,124,234.5 |      - |
|     PR | Plain(...)ookie [26] | 168.58 ns |  5,932,007.3 |  -3.1% |
|   main |           LiveAspNet | 359.96 ns |  2,778,117.3 |      - |
|     PR |           LiveAspNet | 354.01 ns |  2,824,802.3 |  +1.6% |
|   main |               Common | 766.48 ns |  1,304,672.7 |      - |
|     PR |               Common | 721.08 ns |  1,386,812.9 |  +6.3% |

### CODE REVIEW

**Comment by benaadams** [comment]
| Branch |                 Type |      Mean |         Op/s |  Delta |
|------- |--------------------- |----------:|-------------:|-------:|
|   main | TechEmpowerPlaintext |  56.79 ns | 17,609,767.6 |      - |
|     PR | TechEmpowerPlaintext |  53.53 ns | 18,679,950.2 |  +6.1% |
|   main |     PlaintextChunked |  61.89 ns | 16,158,595.4 |      - |
|     PR |     PlaintextChunked |  47.82 ns | 20,912,547.5 | +29.4% |
|   main |  PlaintextWithCookie | 133.94 ns |  7,466,115.3 |      - |
|     PR |  PlaintextWithCookie | 126.26 ns |  7,920,045.3 |  +6.1% |
|   main | Plain(...)ookie [26] | 163.29 ns |  6,124,234.5 |      - |
|     PR | Plain(...)ookie [26] | 168.58 ns |  5,932,007.3 |  -3.1% |
|   main |           LiveAspNet | 359.96 ns |  2,778,117.3 |      - |
|     PR |           LiveAspNet | 354.01 ns |  2,824,802.3 |  +1.6% |

**Comment by benaadams** [comment]
Looks like CI has its own issues

**Review by JamesNK — COMMENTED** [review]
CopyToFast is used by HTTP/1.1, but HTTP/2 uses the generated `Enumerator` types when encoding HPack response headers.

Can this improvement be applied to how the generated `Enumerator` types decide next? Writing response headers in HTTP/2 is a hotspot in our current implementation. Multiplexing and HPack means that only one response can write headers at a time while other responses wait on a lock. Reducing the time in the lock is huge.

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/BadHttpRequestTests.cs [hunk 1]
```diff
@@ -236,9 +236,9 @@ private async Task ReceiveBadRequestResponse(InMemoryConnection connection, stri
             var lines = new[]
             {
                 $"HTTP/1.1 {expectedResponseStatusCode}",
+                "Content-Length: 0",
                 "Connection: close",
                 $"Date: {expectedDateHeaderValue}",
-                "Content-Length: 0",
                 expectedAllowHeader,
```
[the remark this answers is not in the document]
Comment (benaadams) [reply src/Servers/Kestrel/test/InMemory.FunctionalTests/BadHttpRequestTests.cs:242]: Yes

**Comment by benaadams** [comment]
> Can this improvement be applied to how the generated `Enumerator` types decide next?

Yes, though would do it in follow up

**Comment by benaadams** [comment]
Rebased for merge conflict

**Comment by JamesNK** [comment]
> > Can this improvement be applied to how the generated `Enumerator` types decide next?
> 
> Yes, though would do it in follow up

https://github.com/dotnet/aspnetcore/issues/33352

**Comment by benaadams** [comment]
Rebased to retrigger glitchy CI

**Comment by JamesNK** [comment]
I've rerun the build a couple of times and it is consistently failing. The Windows build is timing out.

**Comment by BrennanConroy** [comment]
>D:\workspace\_work\1\s\eng\targets\CSharp.Common.targets(115,5): error MSB5021: Terminating the task executable "cmd" and its child processes because the build was canceled. [D:\workspace\_work\1\s\src\Servers\Kestrel\perf\Microbenchmarks\Microsoft.AspNetCore.Server.Kestrel.Microbenchmarks.csproj]
SUCCESS: The process with PID 5864 (child process of PID 8004) has been terminated.
SUCCESS: The process with PID 8004 (child process of PID 2648) has been terminated.

Microbenchmarks validation is likely hanging. Something changed to make a single run of one or more of the the benchmarks hang.

**Comment by Tratcher** [comment]
In hung_dotnet.exe_2596_210608_031244.dmp I see a thread hung at ConsoleLifetime.OnProcessExit waiting for the host be be disposed.
https://github.com/dotnet/runtime/blob/6f7f30e118d6652a89dd7261e31d33ecd75f66fe/src/libraries/Microsoft.Extensions.Hosting/src/Internal/ConsoleLifetime.cs#L79-L86

**Comment by Tratcher** [comment]
Rebased, dropped the submodule changes, and added a commit that should help avoid the hang.

**Review by Tratcher — APPROVED** [review]

**Comment by BrennanConroy** [comment]
Any idea why the hang started showing up with this change?

**Comment by Tratcher** [comment]
> Any idea why the hang started showing up with this change?

No, it' wasn't even in an affected benchmark.

**Comment by Tratcher** [comment]
Nevermind, I found it:

The benchmark is throwing an exception related to this change (Content-Length ordering changed), and then GlobalCleanup doesn't run.
```
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> System.AggregateException: One or more errors occurred. (Invalid response
Expected:
HTTP/1.1 200 OK
Date: Fri, 02 Mar 2018 18:37:05 GMT
Content-Type: text/plain
Server: Kestrel
Content-Length: 13

Hello, World!
Actual:
HTTP/1.1 200 OK
Content-Length: 13
Content-Type: text/plain
Date: Tue, 29 Jun 2021 22:01:42 GMT
Server: Kestrel

Hello, World!)
 ---> System.InvalidOperationException: Invalid response
Expected:
HTTP/1.1 200 OK
Date: Fri, 02 Mar 2018 18:37:05 GMT
Content-Type: text/plain
Server: Kestrel
Content-Length: 13

Hello, World!
Actual:
HTTP/1.1 200 OK
Content-Length: 13
Content-Type: text/plain
Date: Tue, 29 Jun 2021 22:01:42 GMT
Server: Kestrel

Hello, World!
   at Microsoft.AspNetCore.Server.Kestrel.Microbenchmarks.InMemoryTransportBenchmark.ValidateResponseAsync(Byte[] request, String expectedResponse) in D:\github\AspNetCore\src\Servers\Kestrel\perf\Microbenchmarks\InMemoryTransportBenchmark.cs:line 79
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions) in System.Private.CoreLib.dll:token 0x6002df0+0x11
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken) in System.Private.CoreLib.dll:token 0x6002e0c+0x3f
   at Microsoft.AspNetCore.Server.Kestrel.Microbenchmarks.InMemoryTransportBenchmark.GlobalSetupPlaintext() in D:\github\AspNetCore\src\Servers\Kestrel\perf\Microbenchmarks\InMemoryTransportBenchmark.cs:line 62
   at BenchmarkDotNet.Engines.EngineFactory.CreateReadyToRun(EngineParameters engineParameters) in BenchmarkDotNet.dll:token 0x60009a7+0x0
   at BenchmarkDotNet.Autogenerated.Runnable_1.Run(BenchmarkCase benchmarkCase, IHost host) in RefEmit_InMemoryManifestModule:token 0x600002a+0x26
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Span`1& arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture) in System.Private.CoreLib.dll:token 0x6004f5c+0x23
   at BenchmarkDotNet.Toolchains.InProcess.Emit.Implementation.RunnableProgram.Run(BenchmarkId benchmarkId, Assembly partitionAssembly, BenchmarkCase benchmarkCase, IHost host) in BenchmarkDotNet.dll:token 0x600030b+0x0
```

**Comment by JamesNK** [comment]
Is there a way to make benchmark errors easier to debug? A friendly error message would be much better than an unidentified freeze.

**Comment by BrennanConroy** [comment]
I'm not sure if there is, but in most cases it's easy enough to run the benchmark in VS and see the issue

**Review by JamesNK — APPROVED** [review]

**Comment by Tratcher** [comment]
Thanks @benaadams

**Pull request merged**